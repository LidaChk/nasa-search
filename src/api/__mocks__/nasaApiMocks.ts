import { NasaApiResponse } from '../nasaApi';

export const HTTP_STATUS_MESSAGES: Record<number, string> = {
  200: 'OK',
  400: 'Bad Request',
  404: 'Not Found',
  500: 'Internal Server Error',
  502: 'Bad Gateway',
  503: 'Service Unavailable',
  504: 'Gateway Timeout',
};

export const mockResponseData = (status: number, data?: NasaApiResponse) => ({
  status,
  statusText: HTTP_STATUS_MESSAGES[status] || 'Unknown Status',
  ok: status >= 200 && status < 300,
  data,
  json: async () => data,
});

export const mockNASAData: NasaApiResponse = {
  collection: {
    items: [
      {
        data: [
          {
            nasa_id: 'as11-40-5874',
            title: 'Apollo 11 Mission Image',
            description: "Astronaut's boot print in lunar soil",
            date_created: '1969-07-20T00:00:00Z',
            media_type: 'image',
            keywords: ['Apollo 11', 'Moon', 'Lunar Surface'],
          },
        ],
        links: [
          {
            href: 'https://images-assets.nasa.gov/image/as11-40-5874/as11-40-5874~thumb.jpg',
            rel: 'preview',
            render: 'image',
          },
        ],
      },
      {
        data: [
          {
            nasa_id: 'as11-40-5875',
            title: 'Saturn V Launch',
            description: 'Saturn V rocket lifting off',
            date_created: '1969-07-16T00:00:00Z',
            media_type: 'image',
            keywords: ['Apollo 11', 'Saturn V', 'Launch'],
          },
        ],
        links: [
          {
            href: 'https://images-assets.nasa.gov/image/as11-40-5875/as11-40-5875~thumb.jpg',
            rel: 'preview',
          },
        ],
      },
    ],
  },
};
