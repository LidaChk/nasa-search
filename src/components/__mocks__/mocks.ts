export const mockItem = {
  nasaId: 'test-nasa-id',
  title: 'Apollo 11 Mission Image',
  description: 'Test Description',
  dateCreated: '2024-03-20T00:00:00Z',
  keywords: ['moon', 'space'],
  preview: new URL('https://example.com/preview.jpg'),
  href: new URL('https://example.com/image.jpg'),
};
