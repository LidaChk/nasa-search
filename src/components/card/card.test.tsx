import React from 'react';
import { render, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router';
import Card from './card';
import { mockItem } from '../../__tests__/__mocks__/mocks';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

const mockStore = configureStore({
  reducer: {
    nasaApi: (
      state = {
        queries: {},
        mutations: {},
        provided: {},
        subscriptions: {},
        config: {},
      }
    ) => state,
    selectedItems: (state = {}) => state,
  },
  preloadedState: {
    nasaApi: {
      queries: {},
      mutations: {},
      provided: {},
      subscriptions: {},
      config: {},
    },
    selectedItems: {},
  },
});

jest.mock('react-router', () => ({
  ...jest.requireActual('react-router'),
  useSearchParams: () => [
    new URLSearchParams({
      q: 'moon',
      page: '1',
      details: 'test-nasa-id',
    }),
  ],
  useNavigate: jest.fn(),
}));

const BrowserRouterComponent = (): React.JSX.Element => {
  return (
    <Provider store={mockStore}>
      <BrowserRouter>
        <Card {...mockItem} />
      </BrowserRouter>
    </Provider>
  );
};

const renderCard = () => {
  return render(<BrowserRouterComponent />);
};

describe('Card Component', () => {
  it('renders card with correct content', () => {
    renderCard();

    expect(screen.getByText(mockItem.title)).toBeInTheDocument();
    expect(screen.getByText(mockItem.description)).toBeInTheDocument();
    const image = screen.getByAltText(mockItem.title);
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute('src', mockItem.preview);
  });

  it('renders link with correct navigation path', () => {
    renderCard();

    const link = screen.getByRole('link');
    expect(link).toHaveAttribute(
      'href',
      expect.stringContaining(`details=${mockItem.nasaId}`)
    );
  });

  it('applies correct CSS classes', () => {
    renderCard();

    expect(screen.getByRole('link')).toHaveClass('card');
    expect(screen.getByRole('img').parentElement).toHaveClass(
      'card__image-container'
    );
    expect(screen.getByRole('img')).toHaveClass('card__image');
    expect(screen.getByText(mockItem.title).parentElement).toHaveClass(
      'card__content'
    );
  });

  it('matches the snapshot', () => {
    const { asFragment } = renderCard();
    expect(asFragment()).toMatchSnapshot();
  });
});
