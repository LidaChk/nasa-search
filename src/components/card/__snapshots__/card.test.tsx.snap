// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component matches the snapshot 1`] = `
<DocumentFragment>
  <a
    class="card"
    data-discover="true"
    href="/search?q=moon&page=1&details=test-nasa-id"
  >
    <input
      class="checkbox-input"
      id="test-nasa-id"
      type="checkbox"
    />
    <div
      class="card__image-container"
    >
      <img
        alt="Apollo 11 Mission Image"
        class="card__image"
        src="https://example.com/preview.jpg"
      />
    </div>
    <div
      class="card__content"
    >
      <h3
        class="card__title"
      >
        Apollo 11 Mission Image
      </h3>
      <p
        class="card__description"
      >
        Test Description
      </p>
    </div>
  </a>
</DocumentFragment>
`;
